## Used Car Listings on eBay Kleinanzeigen

In this project we are working with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The data originally uploaded to Kaggle was not available anymore. The [current verion of dataset](https://data.world/data-society/used-cars-data) was provided by Dataquest including immitating a less cleaned version of data.

In [1]:
import pandas as pd
import numpy as np

First, we try to read the CSV file into pandas.

In [2]:
autos = pd.read_csv('autos.csv',encoding = 'Latin-1')

Trying to render a few lines of the autos dataframe:

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

There are 20 columns in our dataset, most of which are of string datatypes. Some columns have null values but all are within 20% of the listings. There are columns containing dates stored as strings.

### Cleaning Column names

First we look into the column names

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

Here we are making two changes:
* Change the camelcase names to snakecase
* Change the wordings of a few columns to more accurately describe the data



In [6]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

### Data Exploration and initial cleaning

In [7]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-05 16:57:05  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

We note the following after initial investigation:
* The columns `seller` and `offer_type` need to be dropped as they both have mostly one value. 
* The `nr_of_pictures` column seems odd and needs more investigation.
* The `price` and `odometer` column are stored as text and need to be cleaned.

In [8]:
autos['nr_of_pictures'].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

As shown above, all values in the `nr_of_pictures` column is 0. Now we move on to convert the data in `price` and `odometer` to numerical values. 

First, the dollar sign and comma are removed from the `price` column. The datatype is also being changed to integer.

In [9]:
autos['price'] = (autos['price'].str.replace(',','')
                  .str.replace('$','')
                  .astype(int)
                 )
autos['price'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

Next, the text "km" and comma is removed from text strings in the `odometer` column. Integer type has been changed to integer, and the column label is changed to `odometer_km`.

In [10]:
autos['odometer'] = (autos['odometer']
                     .str.replace('km','')
                     .str.replace(',','')
                     .astype(int)
                    )
autos.rename({'odometer':'odometer_km'}, axis=1, inplace=True)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

### Exploring odometer and price

First, we look at the data table of odometer numbers:

In [11]:
autos['odometer_km'].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

The odometer numbers are rounded, and most of the cars have high mileage.

In [12]:
print(autos["price"].unique().shape)
print(autos["price"].describe())
autos["price"].value_counts().head(20)

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price, dtype: int64

Findings about the price:
* The price variable has many unique values. 
* 1421 listings have a \$0 price tag - we can remove these rows from our data given they are just 2\% of total number of listings.
* The maximum price is $100M, which needs further investigation

In [13]:
autos['price'].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

A few cars ar epriced at \\$1M or even more. We will remove those that are higher than \\$350000 because the price jumps after going beyond this price.

In [14]:
autos['price'].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

Many cars are priced below \\$30. Given ebay is an auction website, starting bid price of \\$1 could be legitimate. We will be keeping the cars priced at \\$1 and above.

In [15]:
autos = autos[autos['price'].between(1,350000)]
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

Checking the highest and lowest priced cars again:

In [16]:
autos['price'].value_counts().sort_index(ascending=False).head(10)

350000    1
345000    1
299000    1
295000    1
265000    1
259000    1
250000    1
220000    1
198000    1
197000    1
Name: price, dtype: int64

In [17]:
autos['price'].value_counts().sort_index(ascending=True).head(10)

1     156
2       3
3       1
5       2
8       1
9       1
10      7
11      2
12      3
13      2
Name: price, dtype: int64

The data is positively skewed by a few listings that are very expensive. Mean of around \\$6000 is almost double of the median at \\$3000.

### Date columns

There are five columns of data that describe the date:

* `date_crawled`
* `last_seen`
* `ad_created`
* `registration_month`
* `registration_year`

Below, we study the distribution of `date_crawled` data with a data table with percentages.

In [21]:
(autos['date_crawled']
         .str[:10]
         .value_counts(normalize=True, dropna=False)
         .sort_index()
        )

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

The data is fairly uniformly distributed. It looks like the site was crawled over a one month period in 2016 March and April.

In [22]:
(autos['ad_created']
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index())

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

There are many different ad created dates, most heavily concentrated during the data crawling period. There are a few listings that are a bit older, with the oldest being 10 months ago.

In [24]:
(autos['last_seen']
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index())

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The crawler recorded the `last_seen` date it last saw a listing. Supposedly it could determine the date the listing was removed due to the car being sold.

However, almost half of the data was concentrated on the last three days. As it is unlikely for sales to increase 5-10x from previous days, the concentration of data seems to be related to the ending of the crawling period instead of related to sales date.

In [25]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

This is the registration year of the car useful for calculating the age of the listing. The odd minimum and maximum value of 1000 and 9999 means some cleaning is required

### Cleaning registration year column

Since the data was crawled in 2016, any registration year after 2016 is incorrect. On the other hand, the earliest possible registration date for a car could be early 1900s. Let's check the percentage of listings that falls within 1900-2016:

In [32]:
(autos['registration_year'].between(1900,2016)).sum()/autos.shape[0]

0.961206630289303

Since only around 4% of listings are outside this range, let's remove the listings outside of 1900-2016

In [42]:
autos = autos[autos['registration_year'].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

In [39]:
autos["registration_year"].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

More than half listed cars were registered after 2003; the oldest car was registered in 1910.

### Exploring `brand` column

In [45]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

The top few brands are all German brands. To study a more diverse population of cars, the top 20 brands are selected for aggregation.

In [49]:
top_20_brands = autos['brand'].value_counts().head(20).index
print (top_20_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos', 'volvo', 'mini'],
      dtype='object')


In [56]:
brand_mean_prices = {}

for b in top_20_brands:
    mean_price = autos.loc[autos['brand']==b,'price'].mean()
    brand_mean_prices[b] = int(mean_price)
    
brand_mean_prices

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749,
 'renault': 2474,
 'peugeot': 3094,
 'fiat': 2813,
 'seat': 4397,
 'skoda': 6368,
 'nissan': 4743,
 'mazda': 4112,
 'smart': 3580,
 'citroen': 3779,
 'toyota': 5167,
 'hyundai': 5365,
 'sonstige_autos': 12338,
 'volvo': 4946,
 'mini': 10613}

Among the most popular brands, bmw, mercedez benz, and audi are more expensive, after more niche brands like sonstige autos and mini. Opel, renault, and fiat are the more affordable options.

### Exploring Mileage

In [63]:
top_brands = autos['brand'].value_counts().head(6).index
top_brand_mean_prices = {}
top_brand_mean_mileage = {}

for b in top_brands:
    mean_price = autos.loc[autos['brand']==b,'price'].mean()
    top_brand_mean_prices[b] = int(mean_price)
    mean_mileage = autos.loc[autos['brand']==b,'odometer_km'].mean()
    top_brand_mean_mileage[b] = int(mean_mileage)

top_bmp_series = pd.Series(top_brand_mean_prices)
top_bmm_series = pd.Series(top_brand_mean_mileage).sort_values(ascending=False)

brand_info = pd.DataFrame(top_bmm_series, columns=['mean_mileage'])
brand_info['mean_price'] = top_bmp_series
brand_info

mean_mileage  mean_price
bmw                  132572        8332
mercedes_benz        130788        8628
opel                 129310        2975
audi                 129157        9336
volkswagen           128707        5402
ford                 124266        3749

The mean mileage do not vary as much as the mean price. Brands with higher mean mileage tend to be more expensive for the top brands.

### Exploring how odometer affect price

In [75]:
mileage = (autos['odometer_km']
                   .value_counts()
                   .sort_index(ascending=False)
                   .index)

mean_price_by_mileage ={}
for m in mileage:
    mean_price = autos.loc[autos['odometer_km']==m,'price'].mean()
    mean_price_by_mileage[m]=int(mean_price)

mean_price_by_mileage_series = pd.Series(mean_price_by_mileage)
mileage_price_df = pd.DataFrame(mean_price_by_mileage_series,columns=['mean_price'])
mileage_price_df

mean_price
150000        3767
125000        6214
100000        8132
90000         8465
80000         9721
70000        10927
60000        12385
50000        13812
40000        15499
30000        16608
20000        18448
10000        20550
5000          8873

We can see that cars with higher mileage are generally cheaper with the exception of 5000 km.  

In [76]:
for b in top_brands:
    
    mean_price_by_mileage_brand ={}
    for m in mileage:
        mean_price = autos.loc[(autos['odometer_km']==m)&(autos['brand']==b),'price'].mean()
        mean_price_by_mileage_brand[m]=int(mean_price)

    mean_price_by_mileage_brand_series = pd.Series(mean_price_by_mileage_brand)
    
    mileage_price_df[b] = mean_price_by_mileage_brand_series
mileage_price_df

mean_price  volkswagen    bmw   opel  mercedes_benz   audi   ford
150000        3767        3483   5687   1871           5706   5650   2188
125000        6214        6027  11401   3454          10021  12398   3803
100000        8132        7978  13353   4477          14062  15165   4634
90000         8465        8739  14486   4076          16019  16142   4111
80000         9721       10722  16261   6016          15941  16471   5784
70000        10927       11472  17420   7678          17147  21327   7375
60000        12385       12659  21188   6780          21598  20350   7117
50000        13812       13296  23556   8769          24063  22786   9670
40000        15499       15678  24586   8926          28272  26383  10480
30000        16608       15213  23596  10384          19138  26527  12660
20000        18448       17286  26795   9317          27863  27085  12394
10000        20550       19246  31102  12485          29478  28339  15708
5000          8873        4325  10101   3779          10668  15972   7947

Mean prices of cars with different mileage and different brands are tabulated above. It is shown that cars with 5000 km mileage are cheaper for all the top brands, which warrants further investigation.

In [79]:
autos[autos['odometer_km']==5000].describe()

price  registration_year     power_ps  odometer_km  \
count     785.000000         785.000000   785.000000        785.0   
mean     8873.515924        1999.979618    89.729936       5000.0   
std     26550.005709          14.995223   146.933228          0.0   
min         1.000000        1910.000000     0.000000       5000.0   
25%       444.000000        1996.000000     0.000000       5000.0   
50%      1600.000000        2001.000000    74.000000       5000.0   
75%      7500.000000        2012.000000   125.000000       5000.0   
max    350000.000000        2016.000000  1998.000000       5000.0   

       registration_month  nr_of_pictures   postal_code  
count          785.000000           785.0    785.000000  
mean             4.038217             0.0  47369.975796  
std              4.021443             0.0  25468.850938  
min              0.000000             0.0   1069.000000  
25%              0.000000             0.0  26789.000000  
50%              3.000000             0.0  47269.000000  
75%              7.000000             0.0  66299.000000  
max             12.000000             0.0  99974.000000

In [81]:
autos[~(autos['odometer_km']==5000)].describe()

price  registration_year      power_ps    odometer_km  \
count   45896.00000       45896.000000  45896.000000   45896.000000   
mean     5928.18738        2002.960890    118.374630  127649.359421   
std      8572.48026           6.965498    185.469109   36910.962243   
min         1.00000        1910.000000      0.000000   10000.000000   
25%      1250.00000        1999.000000     75.000000  125000.000000   
50%      3100.00000        2003.000000    109.000000  150000.000000   
75%      7500.00000        2008.000000    150.000000  150000.000000   
max    345000.00000        2016.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        45896.000000         45896.0  45896.000000  
mean             5.857722             0.0  51161.188208  
std              3.656444             0.0  25755.842875  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30855.000000  
50%              6.000000             0.0  49845.000000  
75%              9.000000             0.0  72108.000000  
max             12.000000             0.0  99998.000000